soccer/gameplay/tactics/coordinated_pass.py

```diff
@@ -175,15 +189,37 @@ def use_chipper(self, value):
     def on_exit_running(self):
         self.remove_subbehavior('receiver')
 
-    def on_enter_kicking(self):
+    # Funciton to find if ball is moving towards receive point
+    def ball_heading_to_receive_point(self):
+        ball_velocity = main.ball().vel;
+        if ball_velocity.mag() > 0.1:
+            ball_velocity_unit = [ball_velocity.x/ball_velocity.mag(), ball_velocity.y/ball_velocity.mag()]
+            path_vector = main.ball().pos - self.receive_point
+            path_unit = [path_vector.x/path_vector.mag(), path_vector.y/path_vector.mag()]
+            cross_product_mag = abs(ball_velocity_unit[0] * path_unit[1] -
+                ball_velocity_unit[1] * path_unit[0])
+            distance = path_vector.mag()
+            if distance < 0.3:
+                return True # If the ball is already close to the receive point but not heading towards it, dont fail
+            if (distance > 0.5 and cross_product_mag > 0.1) or (distance < 0.5 and cross_product_mag > 0.4) or ball_velocity.mag() < 0.05:
+                return False # If ball is too off course or too slow fail the pass
+        return True
+
+    # Funciton that decides if an opponent robot is too close to the kicking robot
+    def opponent_too_close(self):
+        for bot in main.their_robots():
+            distance = (main.ball().pos - bot.pos).mag()
+            if distance < 0.27:
+                return True # If the ball is too close to the kicking robot then fail
+        return False
+
+    def execute_kicking(self):
```

Comment:
What was the reasoning to swap this from on_enter to execute?

Reply:
Originally I was gonna run the new function in this so needed it to be on execute, but then realized I could do it in the transition lambda, so I can just change it back to on enter. This is the same reasoning for the receiving on execute.